flytekit/core/type_engine.py

```diff
@@ -596,6 +596,58 @@ def guess_python_type(self, literal_type: LiteralType) -> Type[list]:
         raise ValueError(f"List transformer cannot reverse {literal_type}")
 
 
+class UnionTransformer(TypeTransformer[T]):
+    """
+    Transformer that handles a univariate typing.Union[T]
+    """
+
+    def __init__(self):
+        super().__init__("Typed Union", typing.Union)
+
+    @staticmethod
+    def get_sub_type(t: Type[T]) -> Type[T]:
+        """
+        Return the generic Type T of the Union
+        """
+        if hasattr(t, "__origin__") and t.__origin__ is typing.Union:  # type: ignore
+            if hasattr(t, "__args__"):
+                return t.__args__  # type: ignore
+        raise ValueError("Only generic univariate typing.Union[T] type is supported.")
+
+    def get_literal_type(self, t: Type[T]) -> Optional[LiteralType]:
+        try:
+            sub_type = [TypeEngine.to_literal_type(v) if v else "" for v in self.get_sub_type(t)]
+            return _type_models.LiteralType(union_type=UnionType(sub_type))
+        except Exception as e:
+            raise ValueError(f"Type of Generic Union type is not supported, {e}")
+
+    def to_literal(self, ctx: FlyteContext, python_val: T, python_type: Type[T], expected: LiteralType) -> Literal:
+        for t in python_type.__args__:
+            try:
+                return TypeEngine.to_literal(ctx, python_val, t, expected)
+            except Exception as e:
+                logger.debug(f"Failed to convert from {python_val} to {t}", e)
+        raise TypeError(f"Cannot convert from {python_val} to {python_type}")
+
+    def to_python_value(self, ctx: FlyteContext, lv: Literal, expected_python_type: Type[T]) -> Optional[typing.Any]:
+        if lv is None:
+            return None
+        st = self.get_sub_type(expected_python_type)
+        for v in st:
+            try:
+                val = TypeEngine.to_python_value(ctx, lv, v)
+                if val:
```

Comment:
`val` could be `None` if this is an Optional. `None` should not indicate failure

Reply:
`val`  will be `None` only when `Literal` is `None`. if `Literal` is None, we will return None in line 634.
So we should return `val` if and only if `val` has value here, otherwise, we should raise an error.